### ISSUE TITLE
HTTPRPC not working when using authentication

### ISSUE THREAD

**Issue by thesau** [RuudBurger/CouchPotatoServer#4854]
I have a seedbox running rtorrent/rutorrent and when trying to set up the connection via httprpc including username and password, it throws the below error:
```
ERROR[          couchpotato.api] Failed doing api request "download.rtorrent.test": Traceback (most recent call last):
  File "/volume1/@appstore/couchpotatoserver-custom/var/CouchPotatoServer/couchpotato/api.py", line 36, in run_handler
    res = api[route](**kwargs)
  File "/volume1/@appstore/couchpotatoserver-custom/var/CouchPotatoServer/couchpotato/core/_base/downloader/main.py", line 180, in _test
    t = self.test()
  File "/volume1/@appstore/couchpotatoserver-custom/var/CouchPotatoServer/couchpotato/core/downloaders/rtorrent_.py", line 91, in test
    if self.connect(True):
  File "/volume1/@appstore/couchpotatoserver-custom/var/CouchPotatoServer/couchpotato/core/downloaders/rtorrent_.py", line 79, in connect
    self.rt._verify_conn()
  File "/usr/local/couchpotatoserver-custom/var/CouchPotatoServer/libs/rtorrent/__init__.py", line 109, in _verify_conn
    assert "system.client_version" in self._get_rpc_methods(), "Required RPC method not available."
  File "/usr/local/couchpotatoserver-custom/var/CouchPotatoServer/libs/rtorrent/__init__.py", line 153, in _get_rpc_methods
    return(self._rpc_methods or self._update_rpc_methods())
  File "/usr/local/couchpotatoserver-custom/var/CouchPotatoServer/libs/rtorrent/__init__.py", line 142, in _update_rpc_methods
    self._rpc_methods = self._get_conn().system.listMethods()
  File "/usr/local/python/lib/python2.7/xmlrpclib.py", line 1233, in __call__
    return self.__send(self.__name, args)
  File "/usr/local/python/lib/python2.7/xmlrpclib.py", line 1591, in __request
    verbose=self.__verbose
  File "/usr/local/python/lib/python2.7/xmlrpclib.py", line 1273, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/usr/local/couchpotatoserver-custom/var/CouchPotatoServer/libs/rtorrent/lib/xmlrpc/basic_auth.py", line 94, in single_request
    response.msg,
ProtocolError: <ProtocolError for <mysb>.ultraseedbox.com/rutorrent/plugins/httprpc/action.php: 401 Authorization Required>
```
Using exactly the same settings in SickBeard-custom (SickRage) works just fine. Hooking up CP with my (decommisioned) local rutorrent via httprpc without authentication, also works just fine, leading me to think that it has something to do with the way CP does it httprpc authentication.

SickRage is also built on Python and so I thought I would check out the differences between the 2 source codes used for rtorrent. Turns out they are very similar!
However the Rtorrent library in CP is using a method BasicAuthTransport `from rtorrent.lib.xmlrpc.basic_auth import BasicAuthTransport` while SickRage uses RequestsTransport `from rtorrent.lib.xmlrpc.requests_transport import RequestsTransport`

Now here comes the interesting bit: I copied over the SickRage Rtorrent library files over to the CP folder (replacing the original rtorrent library). I then needed to make a few hacks to `rtorrent_.py` because the implementations are slightly different, but I got it to work!! 

Or well at least: the test button in the settings returned 'connection ok'. Unfortunately it did break the actual upload of torrents to rtorrent :( (without any clear error messages so that's where I gave up)

Maybe you could look into how SickRage has implemented the rtorrent connection? They have some nice features such as the option to choose authentication type (none, basic, digest) and whether to require a valid SSL certificate (for those with self signed certificates).

I think you can find their library here: https://github.com/SiCKRAGETV/SickRage/search?utf8=%E2%9C%93&q=rtorrent

**Comment by CouchPotatoBot** [comment]
Thanks for the issue report! Before a real human comes by, please make sure your report has all the below criteria checked

- [ ] Fill in all the dots. They're there for a reason
- [ ] Post logs, either inline (for smaller logs) or using [Pastebin](http://pastebin.com/)
- [ ] Post full version information

Please make sure you also read [contribution guide](https://github.com/RuudBurger/CouchPotatoServer/blob/develop/contributing.md#issues) and followed all the steps. 
Make the title describe your issue. Having "CP not working" or "I get this bug" for 100 issues, isn't really helpful. My master will close issues if there isn't enough information. On a good day he will tag the issue on close with the reason (like `can't reproduce`), but usually he won't, the lazy asshat.

Sometimes my master seems like a grumpy cat and responds with short answers. This isn't (always) because he hates you, but because he's on mobile or busy fixing bugs. If something isn't clear, please let him know. Maybe he can teach me his awesome ways.

Thanks!

**Comment by thesau** [comment]
I did some more testing and can now confirm the issue is related to CP not supporting HTTP digest authentication. Unfortunately my seedbox webserver requires this and I cannot change that as I don't have root rights. 
Would be great if you could implement the method SickRage uses!!

**Comment by RuudBurger** [comment]
Could you please make this in a Pull Request? It's easier for me to test and merge it this way.
If you already have a working structure you can easily make it a Pull Request here on GitHub.
Thanks!

**Comment by thesau** [comment]
Sorry mate I'm not a programmer so can't help you there

**Comment by kreativedev** [comment]
thesau Any chance on you posting your updated rtorrent library to get digest auth working? I'm having the exact same issue as you.

**Comment by thesau** [comment]
Hi mate, the changes I made only resulted in the connection test to work, but not the actual uploading of torrent files. So it's probably not much use to you?

**Comment by kreativedev** [comment]
Dang. Nah that's OK then. Maybe it's something that will eventually get added. Guess I'll look for an alternative to couch potato for now

**Comment by mano3m** [comment]
Maybe @jkaberg or @fuzeman can help here?

**Comment by gpapaz** [comment]
This been closed means what? Can't be fixed? Is fixed? Won't bother? I have the same issue with my seedbox and I hate to use the blackhole mechanism to bypass this.

**Comment by Keelan** [comment]
Also looking for this,  auth failing due to Digest

**Comment by fuzeman** [comment]
I'll try have a look this week, shouldn't be too hard to get digest authentication working..

**Comment by Keelan** [comment]
@fuzeman sounds great, thanks. I'll be happy to test!

**Comment by gpapaz** [comment]
Nice @fuzeman ! Hope you make it work!

**Comment by fuzeman** [comment]
I've updated the [rtorrent-python](https://github.com/fuzeman/rtorrent-python) library to support digest authentication (via the "requests" transport). Just need to update the downloader to support the new authentication method (will do this tomorrow).

**Comment by thesau** [comment]
Awesome! I did notice recently that the connection to my seedbox on Sickrage is no longer working. Both not on SickRageTV and the new Sickrage github.

**Comment by fuzeman** [comment]
The **[issue/4854](https://github.com/fuzeman/CouchPotatoServer/tree/issue/4854)** branch is ready for testing, just switch to advanced mode and set the "Authentication" option to "Digest". I've also added a few more SSL options ("SSL Verify", "SSL CA Bundle") that will likely be helpful for anyone using secure connections with self-signed certificates.

Let me know if you find any issues :smiley: 

---------------------

**Note:** the *issue/4854* branch is based off the "develop" branch, so there could be some unknown bugs with other features that have been added recently.
